receiver/pluginreceiver/plugin.go

```diff
@@ -0,0 +1,211 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pluginreceiver
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Plugin is a templated pipeline of receivers and processors
+type Plugin struct {
+	Title       string      `yaml:"title,omitempty"`
+	Template    string      `yaml:"template,omitempty"`
+	Version     string      `yaml:"version,omitempty"`
+	Description string      `yaml:"description,omitempty"`
+	Parameters  []Parameter `yaml:"parameters,omitempty"`
+}
+
+// LoadPlugin loads a plugin from a file path
+func LoadPlugin(path string) (*Plugin, error) {
+	cleanPath := filepath.Clean(path)
+	bytes, err := os.ReadFile(cleanPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %w", err)
+	}
+
+	var plugin Plugin
+	if err := yaml.Unmarshal(bytes, &plugin); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal plugin from yaml: %w", err)
+	}
+
+	return &plugin, nil
+}
+
+// RenderComponents renders the plugin's template as a component map
+func (p *Plugin) RenderComponents(values map[string]interface{}) (*ComponentMap, error) {
+	template, err := template.New(p.Title).Parse(p.Template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create plugin template: %w", err)
+	}
+
+	templateValues := p.ApplyDefaults(values)
+
+	var writer bytes.Buffer
+	if err := template.Execute(&writer, templateValues); err != nil {
+		return nil, fmt.Errorf("failed to execute template: %w", err)
+	}
+
+	return unmarshalComponentMap(writer.Bytes())
+}
+
+// ApplyDefaults returns a copy of the values map with parameter defaults applied.
+// If a value is already present in the map, it supercedes the default.
+func (p *Plugin) ApplyDefaults(values map[string]interface{}) map[string]interface{} {
+	result := make(map[string]interface{})
+
+	for _, parameter := range p.Parameters {
+		if parameter.Default == nil {
+			continue
+		}
+
+		result[parameter.Name] = parameter.Default
+	}
+
+	for key, value := range values {
+		result[key] = value
+	}
+
+	return result
+}
+
+// CheckParameters checks the supplied values against the defined parameters of the plugin
+func (p *Plugin) CheckParameters(values map[string]interface{}) error {
+	if err := p.checkDefined(values); err != nil {
+		return fmt.Errorf("definition failure: %w", err)
+	}
+
+	if err := p.checkRequired(values); err != nil {
+		return fmt.Errorf("required failure: %w", err)
+	}
+
+	if err := p.checkType(values); err != nil {
+		return fmt.Errorf("type failure: %w", err)
+	}
+
+	if err := p.checkSupported(values); err != nil {
+		return fmt.Errorf("supported value failure: %w", err)
+	}
+
+	return nil
+}
+
+// checkDefined checks if any of the supplied values are defined by the plugin
+func (p *Plugin) checkDefined(values map[string]interface{}) error {
+	parameterMap := make(map[string]Parameter)
```

Comment:
Looks like we're only using this map as a quick lookup the value should probably be a `struct{}{}` to limit the memory impact since we don't care about the value associated with the key.